bigtable_automation/gcf/gcs.go

```diff
@@ -0,0 +1,191 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package gcf runs a GCF function that triggers in 2 scenarios:
+//
+//  1. completion of prophet-flume job in borg. On triggering it sets up new
+//     cloud BT table, scales up BT cluster (if needed) and starts a dataflow job.
+//  2. completion of BT cache ingestion dataflow job. It scales BT cluster down
+//     (if needed).
+//
+// There are two set of trigger functions defined:
+// - ProdBTImportController
+// - PrivateBTImportController
+// which targets on production imports and private imports. The folder structure
+// are different for the two scenarios.
+// The environment variables for deployments are stored in (prod|private)/*.yaml
+package gcf
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"path/filepath"
+	"strings"
+
+	"cloud.google.com/go/storage"
+	"google.golang.org/api/iterator"
+
+	pb "github.com/datacommonsorg/tools/bigtable_automation/gcf/proto"
+	"github.com/pkg/errors"
+)
+
+// FindFolders find all the folders
+// Note: files in <pathPrefix> are ignored.
+// Returns a list of folder names under pathPrefix.
+func FindFolders(ctx context.Context, c *storage.Client, bucket, pathPrefix string) ([]string, error) {
+	it := c.Bucket(bucket).Objects(ctx, &storage.Query{Prefix: pathPrefix})
+	folders := []string{}
+	for {
+		attrs, err := it.Next()
+		if err == iterator.Done {
+			break
+		}
+		if err != nil {
+			return nil, err
+		}
+		// skip files and the pathPrefix folder itself.
+		if !strings.HasSuffix(attrs.Name, "/") || attrs.Name == pathPrefix {
+			continue
+		}
+		pathSuffix := strings.TrimSuffix(strings.TrimPrefix(attrs.Name, pathPrefix), "/")
+		// kgnore subfolders
+		if len(strings.Split(pathSuffix, "/")) > 1 {
+			continue
+		}
+		log.Printf("FindFolders in  %s got %s", pathPrefix, attrs.Name)
+		folders = append(folders, attrs.Name)
+	}
+	return folders, nil
+}
+
+func FindFiles(ctx context.Context, c *storage.Client, bucket, pathPrefix string) ([]string, error) {
+	it := c.Bucket(bucket).Objects(ctx, &storage.Query{Prefix: pathPrefix})
+	files := []string{}
+	for {
+		attrs, err := it.Next()
+		if err == iterator.Done {
+			break
+		}
+		if err != nil {
+			return nil, err
+		}
+		// skip folders
+		if strings.HasSuffix(attrs.Name, "/") {
+			continue
+		}
+		log.Printf("FindFiles in %s got %s", pathPrefix, attrs.Name)
+		files = append(files, attrs.Name)
+	}
+	return files, nil
+}
+
+func BigStorePath(bucket, path string) string {
+	return filepath.Join("/bigstore", bucket, path)
+}
+
+// pathToDataFolder is the "gs://"-prefixed folder that contains raw data.
+// For the folder structure that is expected, please see,
+// https://docs.datacommons.org/custom_dc/upload_data.html
+func GenerateManifest(ctx context.Context, bucket, pathToDataFolder string) (*pb.DataCommonsManifest, error) {
+	client, err := storage.NewClient(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	importGroupName := "custom"
```

Comment:
can we infer the import group name from trigger? As the folder structure is fixed